### PR TITLE
feat: add rounding logic and scale zero fix parse_decimal to match parse_string_to_decimal_native behavior

### LINKED ISSUE #7355: add rounding logic and scale zero fix in parse_decimal to match parse_string_to_decimal_native behavior
- Existing string to decimal cast uses parse_string_to_decimal_native
- parse_string_to_decimal_native does not have support for e-notation
- parse_string_to_decimal_native does rounding at scale, not truncate
- parse_decimal an existing method has e-notation support and use elsewhere like arrow-csv, arrow-json.
- parse_decimal tests with scale 0, fails as shown below 
```rust
assert_eq!(
            parse_decimal::<Decimal128Type>("123.45", 38, 0)?,
            123_i128
        );

assertion `left == right` failed
Left:  12345
Right: 123

```
 - existing test using parse_decimal native works
 
 ```rust
assert_eq!(
            parse_string_to_decimal_native::<Decimal128Type>("123.45", 0)?,
            123_i128
        );
```

-  pars_decimal truncates as defualt behavior, parse_string_to_decimal_native does rounding, hence read from json, or csv of a decimal has truncated value but result of a cast operation has rounded value.

```rust
assert_eq!(
            parse_decimal::<Decimal128Type>("123.4567891", 38, 5)?,
            12345679_i128
        );

assertion `left == right` failed
  left: 12345678
 right: 12345679

```
vs

```
assert_eq!(
            parse_string_to_decimal_native::<Decimal128Type>("123.4567891", 5)?,
            12345679_i128
        );

works fine..
``` 

This issue was raised as part of investigation to fix this [datafusion issue] (https://github.com/apache/datafusion/issues/10315)

if parse_decimal add rounding logic and fix the issue with scale=0, parse decimal can be used for string to decimal cast, thereby getting the scientific notation along with it.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by himadripal**
# Which issue does this PR close?
 - Closes #7355 
 - related to  https://github.com/apache/datafusion/issues/10315
 - Follows up - #6905 ,based on the (https://github.com/apache/arrow-rs/pull/6905#issuecomment-2590125691)

Few important consideration -

- Existing string to decimal conversion uses `parse_string_to_decimal_native`
- `parse_string_to_decimal_native` does not have support for e-notation
-  `parse_string_to_decimal_native` does rounding at scale, not truncate
-  `parse_decimal` an existing method has e-notation support and use elsewhere
-  #6905 added rounding support in `parse_decimal`
-  moved string to decimal conversion to use `parse_decimal` to get support for e-notation.

This PR is a 2nd one to break up #6905 , this one add rounding logic to parse_decimal to match the behavior in existing  `parse_string_to_decimal_native`. 

Closes #.

# Rationale for this change

At present, string to decimal conversion does not support e-notation, in arrow, `parse_string_to_decimal_native` is called to get generic string to decimal. parse_decimal on the other hand is used from generic parse method and it has e-notation support. This PR is adding rounding and scale 0 handling to match the behavior or `parse_string_to_decimal_native` method. Then we can replace `parse_string_to_decimal_native` call with `parse_decimal`. This way, we will get e-notation support too.
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by himadripal** [comment]
This is on top of #7177

File: arrow-cast/src/parse.rs [hunk 1]
```diff
@@ -850,7 +850,16 @@ fn parse_e_notation<T: DecimalType>(
     }
 
     if exp < 0 {
-        result = result.div_wrapping(base.pow_wrapping(-exp as _));
+        let result_with_scale = result.div_wrapping(base.pow_wrapping(-exp as _));
```
Comment (alamb) [inline arrow-cast/src/parse.rs:853]: does this change the behavior of parsing e notation? If so I didn't see any tests

File: arrow-cast/src/cast/decimal.rs [hunk 2]
```diff
@@ -598,7 +599,20 @@ mod tests {
             0_i128
         );
         assert_eq!(
+            parse_decimal::<Decimal128Type>("0", 38, 0)?,
+            parse_string_to_decimal_native::<Decimal128Type>("0", 0)?,
```
Comment (alamb) [inline arrow-cast/src/cast/decimal.rs:603]: Having the same behavior in these two functions seems like a reasonable change to me

**Review by alamb — COMMENTED** [review]
The title of this PR now says it adds rounding logic and makes the logic consistent which sounds good. However, I don't see a ticket that describes the problem. 

This PR's description still says it 

> Closes https://github.com/apache/datafusion/issues/10315

and I did see code changes in the `e` parsing code but I didn't see any tests 🤔  I don't think we can merge code without tests. 

I am sorry to be so pedantic, but arrow-rs is used by many projects now and so evaluating and minimizing downstream impacts is very important. I am trying to avoid the overhead of dealing with releasing regressions like 
- https://github.com/apache/arrow-rs/issues/7069

And I also apologize for the length between review cycles, but as we have mentioned many times, our review bandwidth is very limited.

File: arrow-cast/src/parse.rs (hunk 1 above)
Comment (himadripal) [reply arrow-cast/src/parse.rs:853]: Yes. I missed the tests while splitting the large PR.

File: arrow-csv/src/reader/mod.rs [hunk 3]
```diff
@@ -1286,7 +1286,7 @@ mod tests {
         assert_eq!("53.002666", lat.value_as_string(1));
         assert_eq!("52.412811", lat.value_as_string(2));
         assert_eq!("51.481583", lat.value_as_string(3));
-        assert_eq!("12.123456", lat.value_as_string(4));
+        assert_eq!("12.123457", lat.value_as_string(4));
```
Comment (himadripal) [inline arrow-csv/src/reader/mod.rs:1289]: @alamb  you can see the behavior change in this test of arrow-csv.

File: arrow-cast/src/cast/decimal.rs (hunk 2 above)
Comment (himadripal) [reply arrow-cast/src/cast/decimal.rs:603]: once we are able to move to using parse_decimal and deprecate parse_string_to_decimal_native , these tests will be changed to assert the value in the message section of the assert.

**Comment by himadripal** [comment]
> he title of this PR now says it adds rounding logic and makes the logic consistent which sounds good. However, I don't see a ticket that describes the problem.
> 
> This PR's description still says it
> 
> > Closes [apache/datafusion#10315](https://github.com/apache/datafusion/issues/10315)

Added an issue in arrow-rs #7355

**Comment by himadripal** [comment]
> and I did see code changes in the `e` parsing code but I didn't see any tests 🤔 I don't think we can merge code without tests.

Added e-notation tests

**Comment by himadripal** [comment]
> I am sorry to be so pedantic, but arrow-rs is used by many projects now and so evaluating and minimizing downstream impacts is very important. I am trying to avoid the overhead of dealing with releasing regressions like
> 
> * [[Regression in 54.0.0]. Decimal cast to smaller precision gives invalid (off-by-one) result in some cases #7069](https://github.com/apache/arrow-rs/issues/7069)
> 
I apologize for making this extra overhead.

**Comment by himadripal** [comment]
> And I also apologize for the length between review cycles, but as we have mentioned many times, our review bandwidth is very limited.

I understand, will keep this in mind in future.

**Review by kazuyukitanimura — COMMENTED** [review]
Thanks @himadripal

File: arrow-cast/src/parse.rs [hunk 4]
```diff
@@ -850,7 +850,16 @@ fn parse_e_notation<T: DecimalType>(
...
+        let result_with_one_scale_up =
+            result.div_wrapping(base.pow_wrapping(-exp.add_wrapping(1) as _));
```
[the remark this answers is not in the document]
Comment (himadripal) [reply arrow-cast/src/parse.rs:855]: `exp` in the `parse_e_notation` method is being overriden couple of times based on which direction the decimal needs to shift and if the original string has fractional in it. ( i.e 1.23e-2 has 2 fractional digits).

before this check, `exp` represents number of digits to be removed or added. In this case, exp = -3 

now, result_with_scale = 12
result_with_one_scale_up=123

to round up or down, we need to capture what is the digit next to last digit in the result, in this case 3. How we get it is  = (123 -  12* 10 )

if rounding_digit >=5 then we add +1 to the result
else result remains intact.

Image 3-31-25 at 5 30 PM
I added a debugging screenshot to help understand it more.

File: arrow-cast/src/parse.rs [hunk 5]
```diff
@@ -986,6 +1009,13 @@ pub fn parse_decimal<T: DecimalType>(
                 "parse decimal overflow ({s})"
             )));
         }
+        if scale == 0 {
+            result = result.div_wrapping(base.pow_wrapping(fractionals as u32))
+        }
+        //add one if >=5
+        if rounding_digit >= 5 {
```
[the remark this answers is not in the document]
Comment (himadripal) [reply arrow-cast/src/parse.rs:1016]: first I figure out what is the rounding_digit - digit which is next to the last digit in the final result, if the value of the rounding_digit is >=5, then we add +1 to round up the result, else it remains same. 

i.e ```rust 
"1265E-4" -> with scale 3 -> 0.127
in scale 3 the number would be 0.126 and rounding digit will be 5, as rounding digit is more than 5, the result becomes 0.127

1264E-4" -> with scale 3 -> 0.126
here rounding_digit is 4, which is less than 5, so no need to add 1. 
```

[the remark this answers is not in the document]
Comment (himadripal) [reply arrow-cast/src/parse.rs:1016]: `>= 5` is deciding where to round up or down.

**Review by comphead — COMMENTED** [review]
Thanks @himadripal I do like tests, perhaps we can also add the tests for negative decimals roundings?

Also tests for very big numbers, or very small would be beneficial. What comes to my mind with help of chatGpt

```
// **Very Large Numbers**
        assert_eq!(round_to_places(1e15, 2), 1e15); // Large integer should remain unchanged
        assert_eq!(round_to_places(9999999999999.987, 2), 10000000000000.00);
        assert_eq!(round_to_places(-1e15, 3), -1e15);

        // **Very Small Numbers (Near Zero)**
        assert_eq!(round_to_places(1e-15, 10), 0.0000000000); // Rounds to zero at precision 10
        assert_eq!(round_to_places(-1e-15, 10), -0.0000000000); // Rounds to zero
        assert_eq!(round_to_places(0.000000000123456, 12), 0.000000000123); // Should retain up to 12 decimal places
        
        // **Extreme Edge Cases**
        assert_eq!(round_to_places(f64::MAX, 2), f64::MAX); // Maximum f64 value should remain the same
        assert_eq!(round_to_places(f64::MIN, 2), f64::MIN); // Minimum f64 value should remain the same
        assert!(round_to_places(f64::NAN, 2).is_nan()); // NaN should remain NaN
        assert_eq!(round_to_places(f64::INFINITY, 2), f64::INFINITY); // Infinity should remain Infinity
        assert_eq!(round_to_places(f64::NEG_INFINITY, 2), f64::NEG_INFINITY); // Negative Infinity should remain unchanged
```

File: arrow-cast/src/parse.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kazuyukitanimura) [reply arrow-cast/src/parse.rs:1016]: Ah ok, makes sense

**Review by kazuyukitanimura — APPROVED** [review]

**Comment by alamb** [comment]
@himadripal  -- I am preparing to create a new release hopefully tomorrow. Can you please address @comphead 's testing suggestions soon so we can get this PR into that release?

File: arrow-cast/src/parse.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (alamb) [reply arrow-cast/src/parse.rs:1016]: Perhaps we could add a comment in the code to explain this point to future readers who may have the same question

**Review by alamb — COMMENTED** [review]
@Omega359 suggested on a call this morning that we also test this PR using the DataFusion tests to see if we can see any differences, including the extended tests

I can try to run these later today if no one has a chance to do so. I have a PR to do this partially here:
- https://github.com/apache/datafusion/pull/15466

File: arrow-cast/src/parse.rs [hunk 6]
```diff
@@ -2780,6 +2827,55 @@ mod tests {
...
+            ("1265E-4", ".127", 3),
+            ("12.345e3", "12345.000", 3),
+            ("1.2345e4", "12345", 0),
+            ("1.2345e3", "1235", 0),
+            ("1.23e-3", "0", 0),
+            ("123e-2", "1", 0),
+            ("-1e-15", "-0.0000000000", 10),
+            ("1e-15", "0.0000000000", 10),
+            ("1e15", "1000000000000000", 2),
+        ];
```
Comment (comphead) [inline arrow-cast/src/parse.rs:2846]: lets also add couple more  tests on negative

```
        // Edge cases
        assert_eq!(round_to_places(-2.5, 0), -3.0); // Negative half rounding up
        assert_eq!(round_to_places(-2.49, 0), -2.0); // Just below rounding negative
        // Whole numbers (should remain unchanged)
        assert_eq!(round_to_places(-100.0, 2), -100.0);
```

Comment (himadripal) [reply arrow-cast/src/parse.rs:2846]: there are existing tests for such scenarios [here](https://github.com/apache/arrow-rs/pull/7179/files#diff-cd568dd408cf11db340aa0259545de48fca7a0c0ac680bb9ff9a34956bab52b1L2537-L2558). Nevertheless,  I'll add these as well.

**Review by comphead — APPROVED** [review]
Thanks @himadripal lgtm
However I agree with @alamb and @Omega359 to have another set of tests running in DataFusion on top of this branch before the merge

**Comment by alamb** [comment]
I restarted the integration test job to get a clean run

**Comment by alamb** [comment]
I am running the `cargo bench --bench parse_decimal` benchmark on this branch to make sure there is no performance impact

I also tested with this branch with datafusion's tests locally and they seem to have worked so far. I am running the extended tests now too

```shell
warning: `parquet` (lib) generated 2 warnings
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.30s
     Running bin/sqllogictests.rs (target/debug/deps/sqllogictests-924fc434a23455ca)
Completed 113 test files in 4 seconds
(venv) andrewlamb@Andrews-MacBook-Pro-2:~/Software/datafusion$
```

**Comment by comphead** [comment]
> My benchmark results seem to show that this PR significantly slows down decimal parsing, anywhere fro 3 to 17% slower than main. I will rerun to see if I can reproduce these numbers
> 
> ```shell
> cargo bench --bench parse_decimal
> ```
> 
> Here are the results I got:
> 
> ```shell
> group                      fix_parse_decimal_for_rounding_scale_zero    main
> -----                      -----------------------------------------    ----
> -.123                      1.10     22.8±0.01ns        ? ?/sec          1.00     20.7±0.03ns        ? ?/sec
> -00.1                      1.06     33.5±0.02ns        ? ?/sec          1.00     31.6±0.08ns        ? ?/sec
> -12.                       1.03     38.6±0.04ns        ? ?/sec          1.00     37.3±0.14ns        ? ?/sec
> -123                       1.07     39.2±0.07ns        ? ?/sec          1.00     36.8±0.16ns        ? ?/sec
> -123.                      1.05     42.1±0.09ns        ? ?/sec          1.00     40.3±0.24ns        ? ?/sec
> -123.1                     1.07     41.1±0.04ns        ? ?/sec          1.00     38.4±0.27ns        ? ?/sec
> -123.123                   1.12     32.5±0.03ns        ? ?/sec          1.00     29.0±0.03ns        ? ?/sec
> -123.1234                  1.14     34.3±0.09ns        ? ?/sec          1.00     30.0±0.02ns        ? ?/sec
> -12345678912345678.1234    1.21     88.8±0.13ns        ? ?/sec          1.00     73.4±0.06ns        ? ?/sec
> -99999999999999999.999     1.19     86.8±0.08ns        ? ?/sec          1.00     72.8±0.08ns        ? ?/sec
> .123                       1.09     21.8±0.03ns        ? ?/sec          1.00     19.9±0.02ns        ? ?/sec
> 0.0000123                  1.14     26.3±0.08ns        ? ?/sec          1.00     23.1±0.03ns        ? ?/sec
> 00.1                       1.08     33.1±0.16ns        ? ?/sec          1.00     30.7±0.06ns        ? ?/sec
> 12.                        1.07     38.4±0.06ns        ? ?/sec          1.00     36.0±0.13ns        ? ?/sec
> 123                        1.10     39.1±0.16ns        ? ?/sec          1.00     35.7±0.09ns        ? ?/sec
> 123.                       1.10     43.3±5.95ns        ? ?/sec          1.00     39.4±0.14ns        ? ?/sec
> 123.1                      1.11     41.2±0.13ns        ? ?/sec          1.00     37.2±0.09ns        ? ?/sec
> 123.123                    1.14     32.3±0.03ns        ? ?/sec          1.00     28.2±0.09ns        ? ?/sec
> 123.1234                   1.17     34.2±0.04ns        ? ?/sec          1.00     29.3±0.04ns        ? ?/sec
> 12345678912345678.1234     1.24     90.0±0.09ns        ? ?/sec          1.00     72.8±0.13ns        ? ?/sec
> 99999999999999999.999      1.22     88.1±0.11ns        ? ?/sec          1.00     72.0±0.07ns        ? ?/sec
> ```

Thanks @alamb for pointing it out. Trying to profile it to get some hot spots

**Comment by himadripal** [comment]
@alamb is there instructions how to get this benchmark result against main branch. I can then make changes and test locally to see if I can improve the time.

**Comment by himadripal** [comment]
> @alamb is there instructions how to get this benchmark result against main branch. I can then make changes and test locally to see if I can improve the time.

nvm, found the instructions 
```
git checkout main

cargo bench --bench parse_time -- --save-baseline main

git checkout feature

cargo bench --bench parse_time -- --baseline main
```

**Comment by comphead** [comment]
image
Looks like the newly added wrapping instructions is coslty

**Comment by comphead** [comment]
Just checked on main 

image

**Comment by himadripal** [comment]
An update on this issue, its taking longer than I expected, 
after I replace `base.pow_wrapping(....)` calls with POW10[...], many of the conversion used in benchmark improved by 45% but some regressed by 5 to 10%. POW10 is array of precomputed power of 10. I used it  to store 39 power of 10 elements for Decimal128 which has max precision 38, It will occupy 2.4KB of memory but reduce the CPY cycle  required calculating power of 10. Although I have not figured out how to precompute power of 10 for Decimal256 which needs 77 elements, so this may be reversed. 
Another place I found to be costly is calculating the rounding_digit, I could not figure out a way to avoid that yet, still investigating. if anyone has any thoughts of this, please share.

**Comment by alamb** [comment]
> @alamb would you please share, how did you get the result in tabular format above?

I am using the `critcmp` tool here: https://github.com/BurntSushi/critcmp

My entire script is https://github.com/alamb/datafusion-benchmarking/blob/main/compare_arrow_branch.sh

**Comment by kazuyukitanimura** [comment]
@himadripal 
Just wondering
```rust
let base = T::Native::usize_as(10);
```
where `T: DecimalType`

Will it be faster if you change the `base` to `i128` or `u128` and `pow` it, then convert to `T`?

**Comment by alamb** [comment]
Is this PR ready to merge?

**Comment by himadripal** [comment]
> Is this PR ready to merge?

No @alamb 
The extra instructions for rounding is causing more than allowed degradation. WIP - improve performance. 
One other option, I'm considering is - do the rounding under a flag if performance does not  improve to acceptable level with rounding instructions.

**Comment by alamb** [comment]
> The extra instructions for rounding is causing more than allowed degradation. WIP - improve performance.


Makes sense -- thank you -- I'll mark it as draft then so we don't accidentally merge it
